python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py

```diff
@@ -122,21 +128,93 @@ def get_runs(graphene_info, filters, cursor=None, limit=None):
 ]
 
 
+class AssetComputeStatus(Enum):
+    NONE = "NONE"
+    UP_TO_DATE = "UP_TO_DATE"
+    OUT_OF_DATE = "OUT_OF_DATE"
+
+
+def get_upstream_changed_by_asset(
+    all_asset_nodes: Mapping[AssetKey, "GrapheneAssetNode"],
+    all_asset_records: Iterable[AssetRecord],
+    requested_asset_keys: KeysView[AssetKey],
+):
+    last_changed: Dict[AssetKey, float] = {}
+    statuses: Dict[AssetKey, Optional[AssetComputeStatus]] = {}
+
+    # First, iterate over all the asset records and build a dict of last materialization times.
+    for asset_record in all_asset_records:
+        asset_key = asset_record.asset_entry.asset_key
+        if asset_record.asset_entry.last_materialization:
+            last_changed[asset_key] = asset_record.asset_entry.last_materialization.timestamp
+
+    # Beginning with the set of assets for which status was requested, recurse up the
+    # dependency graphs and fill in statuses that are still `None`. Fill in the statuses
+    # dict as we go to avoid traversing any part of the graph more then once. Note that for
+    # A -> B -> C, if A is newer than B, both B and C are "old".
+    #
+    # Some assets (eg: source or partitioned assets) are not supported and get fixed statuses
+    # the first time they're visited.
+    #
+    def fill_status_for(key: AssetKey):
+        asset_node = all_asset_nodes[key].external_asset_node
+
+        if not asset_node or not asset_node.op_names:
+            return AssetComputeStatus.UP_TO_DATE  # source assets are always "up to date"
+        if asset_node.partitions_def_data:
+            return AssetComputeStatus.UP_TO_DATE  # partitioned assets are not supported
+        elif not last_changed.get(key):
+            return AssetComputeStatus.NONE  # never materialized assets cannot be stale
+
+        # mark us out-of-date if upstream nodes have newer materialization times
+        for dep in asset_node.dependencies:
+            dep_key = dep.upstream_asset_key
+            dep_node = all_asset_nodes[dep_key].external_asset_node
+            if (
+                last_changed.get(dep_key, 0) > last_changed.get(key)
```

Comment:
looks like mypy is complaining about this line... might need to do `> last_changed.get(key, 0)`

Reply:
Thanks! I think I need to get my editor configured to do this typechecking so it's not caught by Buildkite 🙈